src/System.Linq.Expressions/tests/Dynamic/GetIndexBinderTests.cs

```diff
@@ -0,0 +1,100 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Linq.Expressions;
+using Xunit;
+
+namespace System.Dynamic.Tests
+{
+    public class GetIndexBinderTests
+    {
+        private class MinimumOverrideGetIndexBinder : GetIndexBinder
+        {
+            public MinimumOverrideGetIndexBinder(CallInfo callInfo)
+                : base(callInfo)
+            {
+            }
+
+            public override DynamicMetaObject FallbackGetIndex(
+                DynamicMetaObject target, DynamicMetaObject[] indexes, DynamicMetaObject errorSuggestion)
+            {
+                throw new NotSupportedException();
+            }
+        }
+
+        [Fact]
+        public void ArrayIndexing()
+        {
+            dynamic d = new [] {0, 1, 2, 3};
+            for (int i = 0; i != 4; ++i)
+            {
+                Assert.Equal(i, d[i]);
+            }
+        }
+
+        [Fact]
+        public void ListIndexing()
+        {
+            dynamic d = new List<int> {0, 1, 2, 3};
+            for (int i = 0; i != 4; ++i)
+            {
+                Assert.Equal(i, d[i]);
+            }
+        }
+
+        [Fact]
+        public void IndexErrorIsNotBindingError()
+        {
+            dynamic d = new[] {0, 1, 2, 3};
+            Assert.Throws<IndexOutOfRangeException>(() => d[9]);
+            d = new List<int> {0, 1, 2, 3};
+            Assert.Throws<ArgumentOutOfRangeException>("index", () => d[9]);
+        }
+
+        [Fact]
+        public void NullCallInfo()
+        {
+            Assert.Throws<ArgumentNullException>("callInfo", () => new MinimumOverrideGetIndexBinder(null));
+        }
+
+        [Fact]
+        public void CallInfoStored()
+        {
+            CallInfo info = new CallInfo(0);
+            Assert.Same(info, new MinimumOverrideGetIndexBinder(info).CallInfo);
+        }
+
+        [Fact]
+        public void ReturnTypeObject()
+        {
+            Assert.Equal(typeof(object), new MinimumOverrideGetIndexBinder(new CallInfo(0)).ReturnType);
+        }
+
+        [Fact]
+        public void NullTarget()
+        {
+            var binder = new MinimumOverrideGetIndexBinder(new CallInfo(0));
+            var arg = new DynamicMetaObject(Expression.Parameter(typeof(object), null), BindingRestrictions.Empty);
+            Assert.Throws<ArgumentNullException>("target", () => binder.Bind(null, new[] {arg}));
+        }
+
+        [Fact]
+        public void NullArgs()
+        {
+            var binder = new MinimumOverrideGetIndexBinder(new CallInfo(0));
+            var target = new DynamicMetaObject(Expression.Parameter(typeof(object), null), BindingRestrictions.Empty);
+            Assert.Throws<ArgumentNullException>("args", () => binder.Bind(target, null));
```

Comment:
There isn't an exception if `args` is non-null and empty. I think that's *probably* correct in that it would lead to the same error if one tried to access an indexer with too many or too few indices, but I thought I'd just point this out here in case my reasoning is wrong.